locust/web.py

```diff
@@ -477,4 +523,67 @@ def update_template_args(self):
             "stats_history_enabled": options and options.stats_history_enabled,
             "tasks": dumps({}),
             "extra_options": extra_options,
+            "show_userclass_picker": self.userclass_picker_is_active,
+            "available_user_classes": available_user_classes,
+            "available_shape_classes": available_shape_classes,
         }
+
+    def update_environment(self, user_classes, shape_class_name, parsed_options):
```

Comment:
Is this even needed now? Arent we just loading all the files and only changing Users and shape class at run time?

Reply:
I'm pretty confident that it is, because if multiple locustfiles are used, then `main.py` doesn't initiate the Environment with the UserClasses (see line 72 of `main.py`). So `update_environment` needs to update the `user_classes` and `shape_class` properties. The other ones can probably be removed, I forgot to test that out, but will do so now.

Additionally, the application breaks without this line `environment.runner._users_dispatcher = None`

Reply:
> if multiple locustfiles are used, then main.py doesn't initiate the Environment with the UserClasses (see line 72 of main.py).

but why cant main.py initiate it? (Now that we’re not selecting which locustfile to run after startup)

Reply:
That's a good point. I modified `main.py` to:
```python
if userclass_picker_is_active:
    # iterate through each locustfile and 'load' them
else:
    docstring, user_classes, shape_class = load_locustfile(locustfile)
```
I think the `# iterate through each locustfile and 'load' them` block would work for both scenarios. I'll update that.

But i think `update_environment` would still be needed to be called when the user selects a UserClass or Swarm class in the UI. So that the environment is updated with the UserClass/ShapeClass selection and also so that the following is called:
```python
environment.runner.stop()
environment.runner._users_dispatcher = None
```
which causes this user dispatcher to be recreated with the selected UserClasses: https://github.com/locustio/locust/blob/master/locust/runners.py#L491

Reply:
> But i think update_environment would still be needed to be called when the user selects a UserClass or Swarm class in the UI. So that the environment is updated with the UserClass/ShapeClass selection and also so that the following is called

Yes, that makes sense. I’m unsure about having user_picker_is_active as an implied effect of having multiple files specified though? Its a very UI-specific thing but impacts behaviour deep inside main. Maybe have it as a separate setting? Or just always have it on but hide the User selection more in the UI?

Reply:
I think the other change will resolve half of this comment. `main` will essentially only use `userclass_picker_is_active` to toggle the UI to show the UserClass & ShapeClass select fields. The other changes I made using that variable in `main` won't be necessary.

In terms of having `user_picker_is_active` as an implied effect of multiple files, I think the best alternative is to just have it as a CLI Option (which seems like that's what you might be suggesting)

Reply:
👍👍

Reply:
@cyberw I've made all the requested updates:

- Added `--enable-userclass-picker` option to explicitly trigger the UI selections. This cleans up the issues you had with `userclass_picker_is_enabled` behavior in `main`. There are no more conditional checks since `Environment` is always created `user_classes` and `shape_class`.
  - Now `options.enable_userclass_picker` is only used in `main` to enable the UI selections (line 262) and to log that the UserClass picker is enabled (line 445)
- The new workflow is that `/swarm` will update the environment only if the Userclass Picker is enabled AND those values are in the payload. This also lead to removing `update_environment` for a more simplified approach

**There is a breaking change that was implemented with allowing multiple values to be passed into** `-f/--locustfile`. I only just now caught it. Specifying User classes in the CLI cannot be done after the `-f` arguments, since it expects the values to be locustfiles.

Good:
- `locust -f locustfiles --csv test_csv QuickstartUser1`
- `locust QuickstartUser1 -f locustfiles

Bad:
- `locust -f locustfiles QuickstartUser1`
  - Error: `Invalid file 'locustfiles'. File should have '.py' extension`

I didn't update the docs on this yet as I wanted to see if I should keep it as is (maybe at a note in the error message), or perhaps add a flag to specify the User classes.